### ISSUE TITLE
Composition system

### ISSUE THREAD

**Issue by Geokureli** [HaxeFlixel/flixel#3053]
Rather than the current inheritance based system, I think we need to move to a more composition based one, this has been asked a lot, and I think it's doable, without making flixel less accessible to newcomers. I'll be inviting people to this discussion but if you have experience or insight in composition patterns or component systems, feel free to join in, because, unfortunately, I do not.

## Main Goals:
 - Rather than FlxObject being a FlxBasic with collision fields, perhaps it can be a Basic with collision and position components, and we can add those same components to other things like FlxGroups. same with FlxSprite, rather than sprites being a FlxObject with a graphic-related fields and functions it can be a basic with a graphical, position and orientation components, which may or may not have physics/collision.
 - A way to check for and access specific components of any given FlxBasic, maybe something like `obj.has(GraphicComponent)` and a similar `get` method or something, I should probably look into how other frameworks do this.
 - For (somewhat) backwards compatibility we should still have classes for FlxObject and FlxSprite, with the same inheritance and fields that act as delegates for fields and functions moved to components

**Comment by 47rooks** [comment]
Ok, been puzzling over this a bit. The key things I wonder about at the moment are:
    1. do you plan to support adding and removing components from a basic at runtime ?
    2. will all the invocations turn into: `if obj.hasComponent(X) obj.processComponent(X)` and so on or is there a better way ?
    3. can macros make any of this easier and in particular if this is a compile time feature would adding components via metadata help ?

**Comment by 01010111** [comment]
> will all the invocations turn into: if obj.hasComponent(X) obj.processComponent(X) and so on or is there a better way ?

I think this is a cool opportunity to give newcomers the classic flixel experience by using getters and setters to help. So both easing that onboarding developer experience and attempting to make as few breaking changes as possible, a FlxSprite would have getters and equivalent functions that calls whatever `sprite.getComponent()` function it needs under the hood.

**Comment by EyeDaleHim** [comment]
I have a fork of this that tries to add composition to Flixel, it's recent and a work-in-progress, but I think it can be useful for feedback on the main repository - if some form of it gets added.

I am not an expert on general game framework stuff or disciplined enough so I think there might be some mistakes along the way and some indecisiveness on certain things.

[https://github.com/EyeDaleHim/flixel/tree/ecs](https://github.com/EyeDaleHim/flixel/tree/ecs)

**Comment by itulau** [comment]
> I have a fork of this that tries to add composition to Flixel, it's recent and a work-in-progress, but I think it can be useful for feedback on the main repository - if some form of it gets added.
> 
> I am not an expert on general game framework stuff or disciplined enough so I think there might be some mistakes along the way and some indecisiveness on certain things.
> 
> https://github.com/EyeDaleHim/flixel/tree/ecs

Nice work! I like your implementation.

Disclaimer: I'm not a maintainer of HaxeFlixel so don't take what I say too seriously, but still if you don't mind me giving some feedback:

- "objects" being a Array is an interesting choice. Have you tried testing if maybe there is some performance gain by implementing it as a [haxe.ds.ObjectMap](https://haxe.org/manual/std-Map.html) instead with the classes as key? Maybe the tradeoff of more memory (hashtable) for quicker component lookups might be worth it https://try.haxe.org/#52A631E0
- I think "hasComponent" might be of little benefit since we already have "getComponent", which does more or less the same (with a bonus of the reference) and I think the complexity of the functions are practically the same.
- a "getComponents" function like in Unity would be cool to have to if it's allowed for objects to have multiple components of the same type, so a function like this would allow easy access to all of them.
- In my opinion I think unifying position, scale and rotation into one component might be a stronger design choice. These are tightly together if you use transform math to move, scale or rotate this elements. (don't know what @Geokureli thinks about this).
- More controversial territory, but I think Transform component (or the separated things) should be a first class component of FlxObject, again @Geokureli what is your opinion on this?

Other than that, I think your implementation is headed into the right direction.

**Comment by EyeDaleHim** [comment]
Great feedback, I'll just cover one point since I don't have any strong arguments against the rest of them.

> * "objects" being a Array is an interesting choice. Have you tried testing if maybe there is some performance gain by implementing it as a [haxe.ds.ObjectMap](https://haxe.org/manual/std-Map.html) instead with the classes as key? Maybe the tradeoff of more memory (hashtable) for quicker component lookups might be worth it https://try.haxe.org/#52A631E0

If you mean components (I assume in `FlxBasic`) I actually had trouble deciding if I should allow multiple components of one type, if adding a component of the same type implies overwriting, etc etc. I don't know if it's something that should be managed by a HaxeFlixel user, but shedding more light on this might be a big relief.

**Comment by itulau** [comment]
> Great feedback, I'll just cover one point since I don't have any strong arguments against the rest of them.
> 
> > * "objects" being a Array is an interesting choice. Have you tried testing if maybe there is some performance gain by implementing it as a [haxe.ds.ObjectMap](https://haxe.org/manual/std-Map.html) instead with the classes as key? Maybe the tradeoff of more memory (hashtable) for quicker component lookups might be worth it https://try.haxe.org/#52A631E0
> 
> If you mean components (I assume in `FlxBasic`) I actually had trouble deciding if I should allow multiple components of one type, if adding a component of the same type implies overwriting, etc etc. I don't know if it's something that should be managed by a HaxeFlixel user, but shedding more light on this might be a big relief.

Yes, nice catch. I meant the "components" variable you added, instead of "objects". That variable (components) I meant to say for testing its implementation details as an object map. 

Yeah I've thought about what you mention too, it's quite a complicated design choice. Maybe we could follow the trend in this case? I don't know about other EC(S) implementations, but in Unity and Godot (which I've used before), they allow duplicates components respectively, and only ""disallow"" them in certain cases, for whatever reason. Maybe we can do that too?

**Comment by EyeDaleHim** [comment]
> Yeah I've thought about what you mention too, it's quite a complicated design choice. Maybe we could follow the trend in this case? I don't know about other EC(S) implementations, but in Unity and Godot (which I've used before), they allow duplicates components respectively, and only ""disallow"" them in certain cases, for whatever reason. Maybe we can do that too?

I've done a similar thing with an earlier experiment with ECS, you can see that [here](https://github.com/EyeDaleHim/Crow-Engine/blob/dev/source/crow/ecs/components/BaseComponent.hx#L68-L140).

**Comment by itulau** [comment]
> > Yeah I've thought about what you mention too, it's quite a complicated design choice. Maybe we could follow the trend in this case? I don't know about other EC(S) implementations, but in Unity and Godot (which I've used before), they allow duplicates components respectively, and only ""disallow"" them in certain cases, for whatever reason. Maybe we can do that too?
> 
> I've done a similar thing with an earlier experiment with ECS, you can see that [here](https://github.com/EyeDaleHim/Crow-Engine/blob/dev/source/crow/ecs/components/BaseComponent.hx#L68-L140).

Interesting design. I think something like this with a strong default (I'm biased towards Multi) might be a good path. Though I'm not sure if that many choices might complicate the implementations details? Ideally look-ups for components should be as fast as possible if not O(1) using something like macros.

**Comment by EyeDaleHim** [comment]
I went for a Single-type approach since I think it should be simple as much as possible, and to leave headroom as I don't have experience with macros to upkeep performance.

In the meantime, I am interested what else in HaxeFlixel can be transformed to be composition-compliant and how we can do them.

**Comment by ADA-Funni** [comment]
> Rather than the current inheritance based system, I think we need to move to a more composition based one

I was just thinking: maybe you could do that for Flixel 7?

Also, I agree greatly with adding a component system, maybe you could do something like this pseudocode I've written here, inspired by how Unity does component stuff:
```haxe
package;

/*
// FlxState is just this:
var state = new FlxBasic();
state.addComponent(FlxStateComponent);

// state is FlxG.state
// First, you extend FlxBasic and then implement components individually.
*/
class MyState extends flixel.FlxBasic {
    public function new() {
        super();
        addComponent(flixel.components.FlxStateComponent);
        addComponent(flixel.addons.components.FlxStateTransitionComponent);
    }

    override public function create() {
        super.create();
        
        // This is the same as `new FlxSprite()`
        // Adding components should work like `NextState`, but you hook it to a class.
        var sprite = new FlxBasic();
        sprite.addComponent(FlxObjectComponent);
        sprite.addComponent(FlxGraphicComponent);
        sprite.getComponent(FlxGraphicComponent).loadGraphic('assets/images/flixel.png');
        add(sprite);

        // This is the same as `new FlxSprite()`
        // Adding components should work like `NextState`, where you hook it to a function.
        var otherSprite = new FlxBasic();
        otherSprite.addComponent(FlxObjectComponent);
        otherSprite.addComponent(FlxGraphicComponent);
        otherSprite.getComponent(FlxGraphicComponent).loadGraphic('assets/images/flixel.png');
        sprite.add(otherSprite); // You can add sprites to other sprites, just like in Unity!
    }
}
```
Basically, every object can be traced nicely back to FlxBasic, and you can add your own components, which can help with modding, or just general usage of the engine.

**Comment by ADA-Funni** [comment]
I'm done with the pseudocode, I'd really like to see what people think of my design idea! :)

**Comment by Geokureli** [comment]
> I'm done with the pseudocode, I'd really like to see what people think of my design idea! :)

I'm thinking something similar, but I would just add component instances, could retrieve by type but that could be `get<T>(t:Class<T>):T`

**Comment by ADA-Funni** [comment]
> > I'm done with the pseudocode, I'd really like to see what people think of my design idea! :)
> 
> I'm thinking something similar, but I would just add component instances, could retrieve by type but that could be `get<T>(t:Class<T>):T`

Ahh
Maybe you could ID them in a map? Something like `sprite.getComponent("myId")`
